_posts/2020-06-14-dynamic-partition-pruning.md

```diff
@@ -0,0 +1,159 @@
+queries registered with this service can request dynamic filters while scheduling splits without blocking any operations.
+This service is also responsible for ensuring that all the build-side tasks of a join stage have completed execution before 
+constructing dynamic filters for use in the scheduling of probe-side scans from the coordinator.
+
+# Implementation
+
+For identifying opportunities for dynamic filtering in the logical plan, we rely on the implementation added in
+[#91](https://github.com/prestosql/presto/pull/91). Dynamic filters are modeled as `FunctionCall` expressions which 
+evaluate to a boolean value. They are created in the `PredicatePushDown` optimizer rule from the equi-join clauses of inner join 
+nodes and pushed down in the plan along with other predicates. Dynamic filters are added to the plan after the cost-based
+optimization rules to reorder joins and choose join distribution type have been run. This ensures that dynamic filters do not
+interfere in cost estimations and cause a sub-optimal plan to be picked by the CBO.
+The `PredicatePushDown` rule can end up pushing dynamic filters to unsupported places in the plan via inferencing. 
+This was solved by adding the 
+[`RemoveUnsupportedDynamicFilters`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java)
+optimizer rule which is responsible for ensuring that
+* Dynamic filters are present only right above a `TableScan` node and only if the subtree is on the probe side of some downstream `JoinNode`
+* Dynamic filters are removed from `JoinNode` if there is no consumer for it on its probe side subtree.
+
+We also run [`DynamicFiltersChecker`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/sql/planner/sanity/DynamicFiltersChecker.java)
+at the end of the planning phase to verify that dynamic filters from a `JoinNode` are consumed by a `FilterNode` above a `TableScan` node
+on it's probe-side subtree.
+ 
+We reuse the existing [`DynamicFilterSourceOperator`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java)
+in `LocalExecutionPlanner` to collect build-side values from each inner join on each worker node. In addition to passing the collected `TupleDomain`
+to [`LocalDynamicFiltersCollector`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java) 
+within the same worker node for use in broadcast join probe-side scans, we also pass them to `TaskContext` to populate task 
+status updates for the coordinator. 
+
+`ContinuousTaskStatusFetcher` on the coordinator node pulls task status updates from all worker nodes every 
+`task.status-refresh-max-wait` seconds (default is 1 second). `DynamicFilterService` on the coordinator regularly polls 
```

Comment:
> on the coordinator node pulls task status updates from all worker nodes every

This is simplification (see: `io.prestosql.server.TaskResource#getTaskStatus`). Coordinator will fetch task stats at most after 1 second or after task state changes.

Maybe:
```
`ContinuousTaskStatusFetcher` on the coordinator node pulls task status updates from all worker nodes up to every
`task.status-refresh-max-wait` seconds (default is 1 second) or less (if task status changes).
```

Reply:
Updated